bbv1.0

tanyalpd opened the pull request:
# Quality Assurance Checklist
To make reviews more efficient, please make sure the board meets the following standards and check everything off once the board meets the quality check. Once everything has been checked, the assigned reviewers will begin the review process. _Edit this description to check off the list._

There are exceptions with all guidelines. As long as your decisions are justified, then you are good! Contact the reviewers or the leads about any exceptions.

**Please *read* every word on *every* bullet point before checking off the corresponding box.**

## Minimum Prerequisites
- [x] The board passes ERC and DRC.
    - Note some ERC warnings are acceptable (**_CAUTION_** please be careful)
    - Note some DRC warnings are acceptable (less so than ERC: **_CAUTION_** please be careful).
- [ ] All traces are routed.
- [x] Units are in metric.
- [ ] Gerbers are in a zip file in the main directory (after physical design is done)

**Please *read* every word on *every* bullet point before checking off the corresponding box.**

## Schematic Level Requirements 
- [x] Is proper noise resistance given to all peripheral devices (bypass caps and coils/ferrites)?  
- [x] Is proper ESD protection given to all MCU input pins (zener diodes)? You could also write the input pin rating for the MCU on the schematic.  
- [x] Is proper power protection given to peripheral devices (zener diodes)?  
- [x] Are peripheral units used properly (reading datasheet)?  
- [x] Are testing points added at useful places?  
- [ ] Is there proper short to GND protection at MCU outputs (inline resistors)?  
- [x] Do ADC inputs have caps (appropriately sized to prevent RC delay)?  
- [ ] Are ADC inputs biased so there is room above expected value to determine if value is being overflowed?  
- [x] Are LEDs located at useful places (comm, power, debugging, extra GPIO)? They should be placed in locations   
- [x] Are parts chosen easy to collect?  
- [x] Are parts chosen easy to solder?  
- [ ] Is there reverse polarity protection on inputs?  
- [x] Do parts in the schematic match the actual part datasheet?

**Please *read* every word on *every* bullet point before checking off the corresponding box.**

## BOM Requirements
- [ ] **Bill Of Materials**: is there an Excel Sheet BOM and a Mouser Cart link?
- [ ] Are *all* the correct parts in the BOM?
- [ ] Is each part in the Mouser Cart in stock (or do we have the part in stock)?
- [ ] Does each part in the Mouser Cart have a margin for part quantity?
- [ ] Does every component in the schematic have a footprint, mouser link/datasheet, and reference number associated with it?

**Please *read* every word on *every* bullet point before checking off the corresponding box.**

## Layout Level Requirements 
### 2D Spacing
- [ ] The components are spaced out at an optimal distance.
    - All components can be easily hand-soldered.
    - IPC-SM-782A Standard requires a minimum distance of 1.0mm from edge to edge.
- [ ] Components that are in parallel with each other are spaced out at an equal distance when possible.
- [ ] The components are aligned with each other when possible.
- [ ] Components are grouped based off of functionality.
    - E.g. all components for CAN should be grouped.
- [ ] Bypass capacitors are less than 1cm away from their respective IC's power pins.

### 3D Spacing
- [ ] Layout of components have been placed with mounting location and usage of the board in mind.
    - Are PCBs going to be stacked on above/below this board? Are tall components placed accordingly?
    - Are buttons and lights easily accessible and viewable?
    - When mounted into the car, are the heights of the components and connectors accounted for?
- [ ] I am not reading every word and I need to start this checklist over.
    - Please uncheck everything you have checked and start over.
- [ ] Location of connectors and wires going out of the board will prevent messy wiring.
    - Are all the wires that are going in the same direction placed on the same edge of the board?

### Components
- [ ] Do footprints match the schematic?
    - Make sure pins on the footprint correspond to pins in the schematic 
    - (e.g. counterclockwise pins on footprint and alternating pins on schematic is *bad*)
- [ ] Custom footprints have been double checked with the datasheet.
- [ ] Footprints are consistent across the PCB: *i.e. all 0805 for Resistors/Capacitors*
- [ ] Pin 1 of the footprint is labeled in some way. 
- [ ] Are LED's in easy to see places? 
- [ ] Are test points in easy to reach places?  
- [ ] Are critical paths of switching converters as small as possible?  

### Copper Layer 
- [ ] The trace widths and trace clearances are greater than JLCPCB's minimum requirements (https://jlcpcb.com/capabilities/Capabilities). 
    - [ ] Are signal traces are .15mm unless provided with reasoning?   
        - NOTE: One net can have multiple trace widths  
- [ ] The trace lengths are as short as possible. 
    - Can there be a more optimal route if you go to another layer? 
- [ ] Each trace's width is capable of handling the expected current flow. 
    - Use PCB width calculator to calculate trace width. 
- [ ] *No sharp corners. No trace angles equal to or less than 90 degrees. 
    - Orthogonal traces should have vias if necessary. 
- [ ] Are edges of board surrounded by clean ground on both layers with stitching vias?  
- [ ] Traces are in parallel with each other when possible (exceptions can be made to prevent signal coupling). 
    - E.g. traces connected between an IC and MCU can be placed in parallel with each other. 
- [ ] There are no random trace appendages. 
- [ ] Vias placed in copper pads are fully encompassed in the copper pads. 
- [ ] Through-hole components do not have extraneous vias. 
- [ ] Vias should be appropriately sized according to the connected net(s) and expected current capacity.

*Not really a problem for modern manufacturing techniques but good practice and important for high speed signal integrity.

### Silkscreen Layer
- [ ] Visible text sizes are greater than .8mm(L), .8mm(H), .15mm(W).
- [ ] All visible text is on the silkscreen layer.
- [ ] All reference labels of each component are not overlapping a copper pad or another component.
- [ ] All connector pins are labeled with a meaningful and helpful name.
- [ ] All LEDs are labeled with a meaningful and helpful name.
- [ ] Silkscreens are mostly facing one direction (or 2). 

### Edge Cut Layer
- [ ] The dimensions of the board and the mounting holes are nice values in metric i.e. no long decimals.
- [ ] The physical outline of the board is on the edge cut layer.
- [ ] Edge cuts are straight and parallel with opposite edge of the board.
- [ ] Mounting holes are aligned and support M3 screws. All mounting holes should have distances between them shown in comments layer.
- [ ] Corners are curved and contoured to mounting holes.

**Please *read* every word on *every* bullet point before checking off the corresponding box.**

## IMPORTANT NOTICE
- [x] I am confident that this board will be safe to use in a safety critical environment.
- [x] I have read every word on every bullet point and have only checked off a bullet point if I have read every word.
- [x] ((( **OPTIONAL** ))) I had fun :^)

## Other Comments
_Write any comments about the board that would help the reviewers here.

blackbody A - central black body board, communicates to car CAN, i2c line to blackbody c
- controller/CAN communication/power-stepping parts of this board have not been done, only the sensing page which connects to blackbody c

blackbody C - communicates with the thermocouples (embedded in modules) and sensor board, connected to blackbody a, i2c line to other blackbody Cs
 - subject to change from thermocouple reading to thermistor reading, depending on D2s response

sensor board - board that holds the irradiance sensor, on top of the shell, wired to blackbody c

Confluence doc: [https://cloud.wikis.utexas.edu/wiki/spaces/LHRSOLAR/pages/222987986/Blackbody+v.1.0.0](url)

Update 2/25: Added the controller, power distribution, CAN connection to blackbody a based off leader som board
- ERC errors are due to output/input power pins errors

Commit:
Initial commit

Commit:
Git attributes

Commit:
sensor board

Commit:
added more complexity + thermistor connection

Commit:
Initial version of sensor communication through all three boards

Commit:
Updated minor errors

IsaacNeideffer15 commented:
My first comment is can Power Gen start naming things what they are. While blackbody sounds cool, I think it would make it way easier on everyone else if we named it the array sensor boards or something. For the sake of history you could even keep blackbody, but name the project something like "Blackbody: Array sensor boards".

IsaacNeideffer15 commented:
Is there a reason that we are using thermocouplers instead of thermistors? If so I would like to hear about it. Thermocouplers are known to be super susceptible to noise since they output such a tiny voltage. It would be a lot easier and more precise to implement thermistors since you would only need a standard ADC, and would see larger voltage changes with changes in temp. 

I was told by Erika that part of the reason we are using thermocouplers is that they are easy for D2 to manufacture into the array. If that is the only reason then maybe we should talk to D2 about using thermistors.

IsaacNeideffer15 commented:
![image](https://github.com/user-attachments/assets/bacec1e3-76fe-4491-9bd4-ec3af2f65d5f)
You should go ahead and tie your address pin to GND. If you want to do addressing later then you can just mux it. The other option for addressing would be to add some kind of digital potentiometer or DAC, which I wouldn't recommend.

IsaacNeideffer15 commented:
![image](https://github.com/user-attachments/assets/7f8639dc-ca1f-406b-b5cd-132c4d3c2bc1)
What is your reasoning for connecting the alert pins to +3.3v?

I don't think that you should connect the alert pins to something other than ground if you intend to leave them unconnected. They are push-pull so they don't need any kind of pull-up or down resistors if they are connected to the nucleo. I think you should either leave them unconnected or connect them to GND if you don't intend to use them.

IsaacNeideffer15 commented:
![image](https://github.com/user-attachments/assets/20938920-f671-408c-b40b-db4b378499f6)
What is the purpose of the 2M and 2.49M resistors? Are they meant to be pull-up resistors for the ADC input? 

The max input for Vin for the chip you selected is 250 mV so pulling it up that much would be above the safe limit. This can be fixed by changing resistor values. 
![image](https://github.com/user-attachments/assets/66d621ae-d80d-49ba-8911-f115342846e2)

IsaacNeideffer15 commented:
![image](https://github.com/user-attachments/assets/c8346917-9128-46b1-8506-d42616a5e6b3)
What do these jumpers do? Will you put some kind of switches on them so you can enable and disable the chip?

tanyalpd commented:
Thanks for reviewing Isaac! Addressing your points:

- I agree with the renaming, it's something I've been meaning to do but haven't gotten to yet. I'll probably rename blackbody a as ArraySensorBoard and then change blackbody c and the sensor board accordingly

- Thermocouples are currently being used mainly because of what D2 said when we asked about temperature sensing (and because I thought thermocouples were cheaper for some reason when its actually the reverse). We're confirming with D2 about whether its possible to embed thermistors into the modules. I had actually previously done a thermistor schematic already when we were planning to measure ambient temperature and not back-of-the-module so I should be able to reincorporate that.

- Not sure why I should be tying my address pins directly to GND? Because we are planning to have multiple temp sensors, we're pretty much certainly going to be needing multiple addresses. The varying resistor values are meant to allow us to change our address values easily.

- I don't know what happened here but yeah, you're right that they should be unconnected.

- For the thermocouple resistors, I actually read the schematic for MCP9601 instead of MCP9600 when choosing resistor values. However, the resistor values corresponding to MCP9601 are still within that range as seen in the schematic they showed in the datasheet: 
![image](https://github.com/user-attachments/assets/1a64e9bf-570c-45d5-abc5-bfcf7965c105)
![image](https://github.com/user-attachments/assets/5b971a06-d529-4880-b98a-a2a5aae27ebf)
The resistors just provide a voltage divider to allow for open and short circuit detection. I'm not sure if I'm understanding this wrong but from my understanding, I could configure my code to recalibrate the measurement with these resistor values. At the same time though, this fault detection might not be necessary enough to add so I may end up taking the resistors out altogether.

- The jumpers are just solderable ones to enable/disable the chip.

IsaacNeideffer15 commented:
> Thanks for reviewing Isaac! Addressing your points:
> 
> * I agree with the renaming, it's something I've been meaning to do but haven't gotten to yet. I'll probably rename blackbody a as ArraySensorBoard and then change blackbody c and the sensor board accordingly
> * Thermocouples are currently being used mainly because of what D2 said when we asked about temperature sensing (and because I thought thermocouples were cheaper for some reason when its actually the reverse). We're confirming with D2 about whether its possible to embed thermistors into the modules. I had actually previously done a thermistor schematic already when we were planning to measure ambient temperature and not back-of-the-module so I should be able to reincorporate that.
> * Not sure why I should be tying my address pins directly to GND? Because we are planning to have multiple temp sensors, we're pretty much certainly going to be needing multiple addresses. The varying resistor values are meant to allow us to change our address values easily.
> * I don't know what happened here but yeah, you're right that they should be unconnected.
> * For the thermocouple resistors, I actually read the schematic for MCP9601 instead of MCP9600 when choosing resistor values. However, the resistor values corresponding to MCP9601 are still within that range as seen in the schematic they showed in the datasheet:
>   ![image](https://private-user-images.githubusercontent.com/65772998/408425847-1a64e9bf-570c-45d5-abc5-bfcf7965c105.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgzMDUxODYsIm5iZiI6MTczODMwNDg4NiwicGF0aCI6Ii82NTc3Mjk5OC80MDg0MjU4NDctMWE2NGU5YmYtNTcwYy00NWQ1LWFiYzUtYmZjZjc5NjVjMTA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMxVDA2MjgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZmZDY2NzhjNmEwMGRlNWYyMDMyZWE5YjkxNmIyM2IyYmY5ZmM5Y2RiNzZmMzNlNjM5NTNiZGI4NmUzMDc2NmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rBGnHJLhbPNwtVjZGFflsxHz2nbFAnH-V2LcD4i9txg)
>   ![image](https://private-user-images.githubusercontent.com/65772998/408428557-5b971a06-d529-4880-b98a-a2a5aae27ebf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgzMDUxODYsIm5iZiI6MTczODMwNDg4NiwicGF0aCI6Ii82NTc3Mjk5OC80MDg0Mjg1NTctNWI5NzFhMDYtZDUyOS00ODgwLWI5OGEtYTJhNWFhZTI3ZWJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMxVDA2MjgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmNDI0YzM0NzhmYmI1YzFjOWQ0NTgxZGY0OGFlM2FjYTA0NmMwNWE4ODVmZTI4MWNiZGZmZWQ2NWZlNGNjZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0GCs2kDcT1O7zEj0IwFka6RsyftOywGdKaRAUF3LuIo)
>   The resistors just provide a voltage divider to allow for open and short circuit detection. I'm not sure if I'm understanding this wrong but from my understanding, I could configure my code to recalibrate the measurement with these resistor values. At the same time though, this fault detection might not be necessary enough to add so I may end up taking the resistors out altogether.
> * The jumpers are just solderable ones to enable/disable the chip.


Yeah you are definitely right about the addressing sorry I was tweaking. 

The resistors to test for open/closed conditions are still wrong I think. This is what is on the datasheet:
<img width="319" alt="image" src="https://github.com/user-attachments/assets/6502f070-9042-4897-b3e7-02df2cbd0119" />

Yours are 2 and 2.49M.

Commit:
Temperature reader update

KnockbackNemo commented:
Hi Tanya, this looks really good so far!

I didn't review the thermocoupler sheet since it seems like it may be changed (though I still can if it would be helpful!), but on the sensor board, I was looking at the datasheet and wondering why a .1uF capacitor was chosen over a 1uF one. Also, it looks like SDA and SCL may need pull-up resistors? 
![image](https://github.com/user-attachments/assets/4ec50337-70c7-4625-9fa3-49576e44ad97)
![image](https://github.com/user-attachments/assets/0a2fdc03-bfe5-4d69-9ab3-974db6170006)

KnockbackNemo commented:
![image](https://github.com/user-attachments/assets/91cea899-0ead-448e-9e11-ed4445dd3309)

Just curious- how did you choose your resistor values here? I saw that the datasheet recommended using a bunch of equations, and it looks like 4.7k is in the expected range, but I'm not fully sure how to run the math to check.

Other than that it looks good to me!

tanyalpd commented:
Thanks for reviewing my board Maddie!

- Yeah 1uF should be better that was just a mistake
- I put resistors on the scl/sda line on the output of the i2c translator
- the resistor values are variable, but i put an esitmated lower value since I expect there to be higher capacitance due to daisy-chaining (and thus multiple slave devices) and longer wires

KnockbackNemo commented:
Thanks for the explanations!

I saw these pull-up resistors for the main SDA/SCL line, but I think the lines to the sensor board need them too? 
![image](https://github.com/user-attachments/assets/6b338b7a-738e-47b5-aaf6-2b4725ea4c2b)
![image](https://github.com/user-attachments/assets/126f826e-d1c9-49f4-9aec-bc9288d10c95)

tanyalpd commented:
theres pull up resistors on blackbody a where the i2c bus starts. i accidentally put pull up resistors on that same part but i'll switch them to pull up the SCL_board and SDA_board lines instead

KnockbackNemo commented:
Ah I see. Sounds good!

Commit:
Added controller and CAN

Commit:
Added power distribution

Commit:
minor errors

IsaacNeideffer15 commented:
Good work. The blackbody is coming together nicely. Here is what I have noticed so far, I will take another look tommorow. Feel free to tell me im stupid if I'm wrong about anything. 

Blackbody C:
Temperature: 
<img width="250" alt="image" src="https://github.com/user-attachments/assets/b800dbb5-e7ee-4f9d-b6cd-50aa7ca725a0" />
- The datasheet recommends using ferrite beads and diodes like in the image above. 

Blackbody A:
Power Distribution:
<img width="292" alt="image" src="https://github.com/user-attachments/assets/bfb4734f-66e0-4b38-9c61-777e7b7f1c14" />
- The datasheet says that c16 and c17 should be 10uF and 1nF respectively.
<img width="632" alt="image" src="https://github.com/user-attachments/assets/51345364-ea19-4972-bdae-8825f2792235" />
- You should add the 47uF cap from Vout to GND. This is really important because in switching converters like this one they output a rough triangular wave.

Power Protection:
- The datasheet recommends connecting both NCs to RTN for thermal efficiency. This doesn't affect the function of your circuit though. 

Some of your ICs dont have datasheets linked to them. It would be helpful to have them there to make sure we are all looking at the same thing, and it is convenient.

Lakshay983 reviewed (changes requested):
Looks really good!
I see you forked a copy of UTSVT KiCAD libraries. You're supposed to have it as a submodule. A submodule is essentially another repo embedded into this repository. So we can version control the shared library and if someone else makes a change to the part you use (like adding a 3D model), you can pull from the latest commit and get it from there. See the starter project's symbol library sections for how to do that: https://cloud.wikis.utexas.edu/wiki/x/GYDGCw

Also just in general, i'd recommend adding testpoints for things like power, I2C, etc, anything you predict you might need to debug.

Blackbody A:
Also, is the plan for this board to be on the main CarCAN bus on a peripheralCAN bus. A lot of the circuitry is for an isolated CanBus on a peripheral network. 
tldr of the differences is that CarCAN is completely unsiolatd and uses the same GND as the Fusebox and has no isolation. The peripheral bus boards have a layer of isolation on CAN between the fusebox and the rest of the peripheral board.
If you want to make this a board on the CarCAN bus see how the LeaderSOM does it circuitry wise: https://github.com/lhr-solar/LeaderSOMPCB/tree/V1Fixes
If it's a board on the CarCAN bus then you should make it 2 1x2 connector for CAN and 1 1x2 connector to 12V. The peripheral vs leader stuff is kinda confusing so lmk if you want more clarification on it.

Is the plan to use a bare STM32 chip on this board or to use a Nucleo? Either work, but if you're using a bare STM32 chip then you should use a different chip since we have a standardized one. It's the same one as on the PeripheralSOM. Also the one you have on there is in a very annoying solder package.

I wouldn't recommend putting LEDs directly on CAN, the added capacitance may mess with things. Maybe an LED controlled by software indicating that you're getting CAN messages. Also, you can add testpoints for CAN TX, RX, H, and L if you really want to test it

Sensor board:
I'd recommend putting pullups on the sensor board I2C, you may not need it but I think it's good practice to have the pullups there rather than not having it. Also, can the Blackbody A board only support one sensor board? It doesn't look like you can change the address of your irradiance sensor, so you're gonna have hella address conflicts if you try and add multiple sensor boards. You might be able to find an I2C address translator.

Commit:
Fixed pins, minor changes, removed power isolation

Commit:
add testpoints

Commit:
review changes

Commit:
layout iteration 1

Lakshay983 reviewed (changes requested):
pls delete all the backups and add em to ur gitignore

I like the small form factor. 

First big thing I see is that the power traces are pretty undersized. Higher currents require thicker trace lengths: 
So ur 12V_EXT, 12V, 3.3V, and GND need to be thicker than the regular signal wires.
I'd recommend making your 12V_EXT and 12V both planes like you how GND is. 
Also you should make your top layer also GND

The car vibrates a lot, so you can't use pin headers in the car since those will likely shake out, which is why we use molex connectors. Change your CAN_In and CAN_Out connector to be molex 1x2s. Same for the Blackbody C connector, make that a molex 1x4


For the MCU:
The crystal oscillator circuit is very sensitive to noise so you need to minimize the length of those traces on the circuit and also make a keepout zone under those traces.
![image](https://github.com/user-attachments/assets/77bc8f99-4f24-45d8-a2b3-7f656befcb91)

There's a lot of rlly long traces that don't need to be that long: 
![image](https://github.com/user-attachments/assets/20d3b0c9-cfe5-49d8-9047-491df41ab213)
A way to mitigate this is to do layout by subcircuit, it's really hard to conceptualize how to minimize every trace in the board, it's a lot easier to conceptualize it per small circuit block.
So for example, take your power protection circuit: 
![image](https://github.com/user-attachments/assets/ca8e3488-d5d3-4a48-b686-c4d0d10e7861)
Do layout and traces only for that circuit
So smth like this:  
![image](https://github.com/user-attachments/assets/b6217899-31c7-4c85-aacc-e2f437b0ecf0) <- don't do this layout tho, u can make it a lot better
You should do the same thing for CAN and mcu

Make sure your bypass caps for your IC (specifically the mcu) are closer to the pins they're bypassing
It's easier to do bypass cap layout when ur doing subcircuit layout

Don't use a via for the mounting hole, use a mounting hole footprint. The one we use like the m3.2 mounting hole foorpint iirc.

You have a lot of unused pins on the MCU, I'd add some breakouts for extra debugging

The pink thing represents the clip that sticks out represents the mating connector when you plug the female version into it. The clip thingy should be facing inward.
![image](https://github.com/user-attachments/assets/e2dbbd34-7d86-4397-803d-05ed642e0aac)

Overall, great job on the layout!

Commit:
layout edits

Commit:
layout updates

Commit:
fixed footprint issue
